### PR TITLE
Introduced InTheFuture and InThePast specifications.

### PR DESCRIPTION
**Pull request by klesniak1**

### CODE REVIEW

File: src/test/java/net/kaczmarzyk/spring/data/jpa/domain/InTheFutureTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+				.nextSpecialOffer(PAST_DATE_TIME)
+				.birthDate(FUTURE_DATE_TIME.toLocalDate())
+				.lastOrderTime(FUTURE_DATE_TIME.toLocalDateTime().plusMonths(2))
+				.build(em);
+		moeSzyslak = customer("Moe", "Szyslak")
+				.registrationDate(PAST_DATE_TIME.getYear(), PAST_DATE_TIME.getMonth().getValue(), PAST_DATE_TIME.getDayOfMonth())
+				.nextSpecialOffer(FUTURE_DATE_TIME.plusDays(30))
+				.birthDate(FUTURE_DATE_TIME.toLocalDate().plusDays(10))
+				.lastOrderTime(PAST_DATE_TIME.toLocalDateTime())
+				.build(em);
```
Comment (jradlica) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/domain/InTheFutureTest.java:53]: Please verify is there any way to increasing the test data readability.

By adding comment like:
```
/*
*          | registration | next special offer | birth date | last order time
* ------  | ------------ | ------------------ | ---------- | ----------------
* HOMER   |    FUTURE    |       FUTURE       |    PAST    |    FUTURE
* MARGE   |    FUTURE    |       PAST         |   FUTURE   |    FUTURE
* MOE        |    PAST      |      FUTURE        |   FUTURE   |     PAST
*/
```
or by extending the customer builder:
```
		customer("Homer", "Simpson").with(
						REGISTRATION_DATE, IN_THE_FUTURE
						NEXT_SPECIAL_OFFER, IN_THE_FUTURE
						BIRTH_DATE, IN_THE_PAST,
						LAST_ORDER_TIME, FUTURE_DATE_TIME
				).build(em);
```
or
```
		customercustomer("Homer", "Simpson")
                  .withAllDatesInTheFuture()
                  .withBirthDate(IN_THE_PAST).build(em);
```

File: src/main/java/net/kaczmarzyk/spring/data/jpa/domain/InTheFuture.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	private static final long serialVersionUID = 1L;
+
+	public InTheFuture(QueryContext queryContext, String path) {
+		super(queryContext, path);
+	}
+
+	@Override
+	public Predicate toPredicate(Root<T> root, CriteriaQuery<?> query, CriteriaBuilder cb) {
+		Expression<TimeType> rootPath = path(root);
+		Expression<TimeType> currentTimestamp = (Expression<TimeType>) cb.function("CURRENT_TIMESTAMP", rootPath.getClass());
```
Comment (tkaczmarzyk) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/domain/InTheFuture.java:44]: have you verified that this is supported on JPA-level and/or which databases support it?

File: src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java [hunk 3]
```diff
@@ -54,6 +52,7 @@ public class Customer {
 
     @JsonFormat(pattern = "yyyy-MM-dd")
     private Date registrationDate;
+	private Calendar registrationCalendar;
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java:55]: invalid formatting

File: src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java [hunk 4]
```diff
@@ -75,7 +74,9 @@ public class Customer {
     private Boolean goldObj;
     
     private Instant dateOfNextSpecialOfferInstant;
-    private OffsetDateTime dateOfNextSpecialOffer;
+	private Timestamp dateOfNextSpecialOfferTimestamp;
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java:77]: invalid formatting

File: src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java [hunk 5]
```diff
@@ -140,6 +141,9 @@ public void setFirstName(String firstName) {
 
     public void setRegistrationDate(Date registrationDate) {
         this.registrationDate = registrationDate;
+
+	    this.registrationCalendar = Calendar.getInstance();
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java:145]: invalid formatting

**Review by tkaczmarzyk — CHANGES_REQUESTED** [review]

File: src/main/java/net/kaczmarzyk/spring/data/jpa/domain/InThePast.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import net.kaczmarzyk.spring.data.jpa.utils.QueryContext;
+
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import javax.persistence.criteria.Expression;
+
+/**
+ * <p>Filters with ">" where-clause (e.g. {@code where dateOfBirth < localtimestamp()).</p>
```
Comment (tkaczmarzyk) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/domain/InThePast.java:27]: ">" --> "<"

File: src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java [hunk 7]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2014-2022 the original author or authors.
- *
+ * <p>
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java:3]: why did you add it?

Comment (tkaczmarzyk) [reply src/test/java/net/kaczmarzyk/spring/data/jpa/Customer.java:3]: generally the license headers are managed automatically via Maven license plugin

**Review by tkaczmarzyk — CHANGES_REQUESTED** [review]

**Review by tkaczmarzyk — APPROVED** [review]

**Pull request merged**